vm-setup/roles/v1aX_integration_test/tasks/node_reuse.yml

```diff
@@ -5,121 +5,167 @@
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+  - name: Get provisioned BMH before upgrade in KubeadmControlPlane node reuse test scenario.
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+      register: provisioning_bmh
 
-  - name: Update Metal3MachineTemplate nodeReuse field to 'True'.
-    shell: |
-        kubectl get m3mt "{{ CLUSTER_NAME }}"-controlplane -n "{{ NAMESPACE }}" -ojson | jq '.spec.nodeReuse=true' | kubectl apply -f-
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-
-  - name: Upgrade KubeadmControlPlane k8s version from "{{ KUBERNETES_VERSION }}" to "{{ UPGRADED_K8S_VERSION }}" to test KCP scale-in feature.
-    shell: |
-        kubectl get kubeadmcontrolplane "{{ CLUSTER_NAME }}" -n "{{ NAMESPACE }}" -o json | jq '.spec.version="{{ UPGRADED_K8S_VERSION }}"' | kubectl apply -f-
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+  - name: Get provisioned BMH names and UUIDs mapping before upgrade in KubeadmControlPlane node reuse test scenario.
+    set_fact:
+      bmh_before_upgrade_kcp: "{{ bmh_before_upgrade_kcp | default([]) + [item.metadata.name + '=metal3://' + item.metadata.uid] }}"
+    with_items: "{{ provisioning_bmh.resources | filter_provisioning('provisioned') |  json_query('[*]') }}"
+    
+  # Tasks to update maxSurge field in KubeadmControlPlane to 0 to start testing KCP scale-in feature
+  # Upgrade KubeadmControlPlane k8s version from "{{ KUBERNETES_VERSION }}" to "{{ UPGRADED_K8S_VERSION }}" to test KCP scale-in feature
+  - name: get KubeadmControlPlane
+    k8s_info:
+      api_version: v1alpha3
+      kind: KubeadmControlPlane
+      namespace: "{{ NAMESPACE }}"
+      name: "{{ CLUSTER_NAME }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: KubeadmControlPlane
+
+  - name: Save KubeadmControlPlane json file
+    copy:
+      content: "{{ KubeadmControlPlane.resources | first }}"
+      dest: "/tmp/KubeadmControlPlane.json"
+
+  - name: Update maxSurge field in KubeadmControlPlane to 0 and upgrade KubeadmControlPlane k8s version from "{{ KUBERNETES_VERSION }}" to "{{ UPGRADED_K8S_VERSION }}"  in json file
+    json_patch:
+      src: "/tmp/KubeadmControlPlane.json"
+      operations:
+        - op: replace
+          path: "/spec/rolloutStrategy/rollingUpdate/maxSurge"
+          value: 0
+        - op: replace
+          path: "/spec/version"
+          value: "{{ UPGRADED_K8S_VERSION }}"
+
+  - name: Apply maxSurge field and upgrade k8s version on KubeadmControlPlane k8s
+    k8s:
+      state: present
+      src: /tmp/KubeadmControlPlane.json
+      apply: yes
+  # End tasks to update maxSurge field and upgrade k8s version in KubeadmControlPlane to 0 to start testing KCP scale-in feature
+
+  # Tasks to update Metal3MachineTemplate nodeReuse field to 'True'
```

Comment:
I have switched the sequence to: 

1. First update Metal3MachineTemplate nodeReuse field to 'True'
2. Update maxSurge field
3. Upgrade KubeadmControlPlane k8s version

Since I prefer to do update maxSurge and Upgrade KCP in the same task to avoid repeating the same tasks.
This should be ok if it is not necessary to Update maxSurge before updating nodeReuse?

Reply:
Yes, it is fine, updating maxSurge before or after updating nodeReuse field is not important, but we must ensure nodeReuse field update before KCP upgrade.